### PR TITLE
StateSpace Module Hurricane Case Study WIP

### PR DESCRIPTION
**Pull request by Dekermanjian**
# Case Study to showcase the use of the State Space Module in `pymc-experimental`

This relates to #715 

In this case study I have done the following:
- Briefly explain how SSMs work under the hood
- Walk through the simplest form of SSM for tracking a 2-D object

I still wish to add the following:
- Expand on the simplest SSM by adding covariates/exogenous variables
- Implement a VARMAX model in the state space representation
- Add non-linear covariates/exogenous variables like HSGPs

@jessegrabowski Once I have completed this, would you be willing to be a reviewer for this case study?

### CODE REVIEW

**Comment by Dekermanjian** [comment]
<p>Hi @fonnesneck! Thank you for taking the time to review this notebook. On ReviewNB it seems that the double $ that should generate an equation on its own line doesn't render, but it does on the PyMC examples page. That is why there are stray $ characters.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2571463966'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Yes, you are correct. I will move the comment on line 6 to the very top of this cell. I have Plotly included in the extra_installs in the myst.md file. Again this does not render in ReviewNB but it does on the PyMC examples website. You can see it if you visit the documentation preview <a href="https://pymcio--763.org.readthedocs.build/projects/examples/en/763/case_studies/ssm_hurricane_tracking.html" target="_blank">https://pymcio--763.org.readthedocs.build/projects/examples/en/763/case_studies/ssm_hurricane_tracking.html</a></p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2571463968'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Thank you for catching that! Yes, I agree with you. I will add some more context on what constitutes an origination point.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2571463971'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jessegrabowski** [comment]
This is really amazing work! Sorry I slept on this for so long. Structurally I think it's good to go, it just need a bit of polish in a few sections. The only major revision I would consider is not using out-of-sample exogenous data to do the exogenous forecasts, to make it more fair against the kinematics-only model. You could also consider adding a more formal "test" of each model, for example the RMSE of each point the predicted path vs the observed path as a time series. You can compute that sample-wise and plot the mean and HDI for each model against each-other, showing the change in expected error, and the variance in the errors, over time.

Also quite curious why your model is so hostile to Newfoundland :)

**Comment by Dekermanjian** [comment]
<p>Totally agree with all of the above. I will clean up the equations with latex and make a note on the Joseph form of the filtered covariance equation. I will also look up what the convention is for the t subscript and replace it everywhere in the notebook that occurs. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910289'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Thank you for making me aware of the numpy formatting. I will make the change everywhere this applies.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910324'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>I didn't think of that. I will make a note of it in the text and I will play around with different P0 initializations.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910367'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>I will test it out with the latest versions of pymc + pymc-extras and report back if it isn't. I recall this being fixed, though.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910411'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Yes, that is correct. I agree that a legend should be included. I will add one!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910461'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Yes, you are correct. I think I need to clarify that with this model we aren't able to model any angular acceleration whenever the hurricane makes maneuvers so we aim to capture turns with the stochastic innovation on the acceleration term.</p><p> </p><p>I will put in the ODE integrations. I think this will be good for completeness. Of course, I will fix the mismatch between the latex and the text and the typo.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910492'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Yes, I will give it a go. I believe at the time the Jax backends were not implemented yet. I will give that a try too.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910551'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Thank you so much for clarifying this for me!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910594'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>I will go back and reword how the stochastic innovations are described in the text.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910640'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>I agree, that priors are poorly specifies. I will go back and pick more sensible priors and I will condense the output with a forest plot.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910685'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Ah yes, I should've been more careful. I will make the change.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910738'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Yes, totally agree. I will make the change.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910777'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Yeah, I must've been testing different variances for P0. I will get rid of the redundant '*1'.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910819'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Yes, you are absolutely correct. I shouldn't be passing in the real exogenous data from the future. I will extend the last observed value and look into if I can also use lagged variables. I also agree that this section needs to read better and be clearer about what are the conclusions.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910855'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
> This is really amazing work! Sorry I slept on this for so long. Structurally I think it's good to go, it just need a bit of polish in a few sections. The only major revision I would consider is not using out-of-sample exogenous data to do the exogenous forecasts, to make it more fair against the kinematics-only model. You could also consider adding a more formal "test" of each model, for example the RMSE of each point the predicted path vs the observed path as a time series. You can compute that sample-wise and plot the mean and HDI for each model against each-other, showing the change in expected error, and the variance in the errors, over time.
> 

Thank you so much for taking time to review this work!!

I will include your suggestion of a more formal test for each model to fairly compare performance. 

> Also quite curious why your model is so hostile to Newfoundland :)
hahaha! If it makes you feel better, quoting Box "All models are wrong, but some are useful."

**Comment by jessegrabowski** [comment]
<p>It happens here if you're interested:</p><p> </p><p><a href="https://github.com/pymc-devs/pymc-extras/blob/7d62c53139419df8f9d7ef89b69ec07befbbad4d/pymc_extras/statespace/core/statespace.py#L2061" target="_blank">https://github.com/pymc-devs/pymc-extras/blob/7d62c53139419df8f9d7ef89b69ec07befbbad4d/pymc_extras/statespace/core/statespace.py#L2061</a></p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910594'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Hey Jesse, the GitHub issue for this issue is still open <a href="https://github.com/pymc-devs/pymc-extras/issues/424" target="_blank">https://github.com/pymc-devs/pymc-extras/issues/424</a></p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910411'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jessegrabowski** [comment]
> Hey Jesse, the GitHub issue for this issue is still open https://github.com/pymc-devs/pymc-extras/issues/424

Ok, I'll work on that ASAP

**Comment by Dekermanjian** [comment]
<p>Hey Jesse, I think maybe I misunderstood one of your comments. The letter "t" itself isn't the issue in your comment but the fact that I have it subscript t within the text under the definition column of the above tables. Is that correct? </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910289'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
Hey @jessegrabowski, I believe I made all of the changes that you suggested. There is one thing that I can't figure out how to do, though. I want to cite your notebook in [pymc-extras](https://github.com/pymc-devs/pymc-extras/blob/main/notebooks/Making%20a%20Custom%20Statespace%20Model.ipynb) but I can't figure out how to do so following the pymc-examples [guide](https://www.pymc.io/projects/docs/en/latest/contributing/jupyter_style.html#authorship-and-attribution). Do you happen to know how to do this?

**Comment by Dekermanjian** [comment]
Hey @jessegrabowski, I re-ran the notebook but something is not working correctly with the exogenous forecasts. While the previous time dimension error message is gone, the output of the forecasts don't look correct. I am trying to write a very simple example of a custom state space model to see if I can debug/figure out what is happening.

**Comment by AlexAndorra** [comment]
I'm guessing [you're running into this](https://github.com/pymc-devs/pymc-extras/issues/491) @Dekermanjian, aren't you?

**Comment by Dekermanjian** [comment]
Hey @AlexAndorra, yes I believe that is the same issue. In my case the output is much worse, however, probably because my ‘k_endog=2’. I am going to try to track down the issue.

**Comment by AlexAndorra** [comment]
Yeah that sounds about right. It'd be hugely helpful if you managed to track this down! LMK if there anything I can do to help

**Comment by jessegrabowski** [comment]
<p>Yes that's correct</p><p> </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2775910289'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jessegrabowski** [comment]
Posted a final round of suggestions. It's mostly nitpicks, so please feel free to ignore many of them. Let's get this merged, it's really amazing!

**Comment by Dekermanjian** [comment]
<p>Hey Jesse, no I didn't cancel it. I've had 2 weird things happen using the nutpie progress bar. First, is the above where the progress bar visual freezes but it continues to sample. Second, is specific to when I ssh into a linux machine using vscode on windows the nutpie progress bar (only nutpie) will print out a bunch of empty lines below the progress bar as it samples.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2972769893'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
<p>Thank you, Jesse for going through this again. I will make the suggested changes and re-push!</p>

**Comment by Dekermanjian** [comment]
<p>Hey <span class="mention" data-value="jessegrabowski" data-denotation-char="@">﻿<span contenteditable="false">@jessegrabowski</span>﻿</span>, is there a way I can get the predicted covariance from <code>sample_conitional_posterior</code>?</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2972769875'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jessegrabowski** [comment]
<p>Gross, ok. As long as it's not real it doesn't matter</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/763/discussion/#comment-2972769893'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by Dekermanjian** [comment]
Hey @OriolAbril, sorry @ you out of the blue but I know you are an expert with how the `pymc-examples` website works and I am having an odd issue where when I check the documentation preview none of the LaTeX math is rendering correctly. Do you have any idea what could possibly be causing this?

**Comment by Dekermanjian** [comment]
Thank you, @OriolAbril! I will try this now!

**Comment by OriolAbril** [comment]
> From the samples -- idata.cov(dim=['chain', 'draw']) ﻿

I don't think this will work. I haven't gone over the notebook to see where this would be used, but if it is related to a question @jessegrabowski asked about covariance matrices in one of the discords you should be able to do something like:

```python
target_variable = idata.posterior["variable_name"]  # has dims (chain, draw, time, state)
corr = xr.corr(target_variable, target_variable.rename(state="state_bis"), dim=("chain", "draw"))
# corr has dims (time, state, state_bis)
```

All elements of corr where the coord for state and state_bis match are 1. Other slices (e.g. `time=time_i, state=state_j, state_bis=state_k`) are the correlation coefficient between `target_variable.sel(time=time_i, state=state_j)` and `target_variable.sel(time=time_i, state=state_k)`. There is also an `xarray.cov` that behaves like corr dimension wise but computes covariances (basically xr.corr->np.corrcoef and xr.cov->np.cov)

**Comment by Dekermanjian** [comment]
@OriolAbril, yeah it might be a plotly issue. I remember in earlier versions when the LaTeX was rendering properly the plotly figures were not rendering. At the time I selected a specific plotly renderer (https://plotly.com/python/renderers/) that rendered properly on NBReview so that the reviews could be done. I will check if there are any renderers that don't use mathjax.

**Comment by Dekermanjian** [comment]
> I don't think this will work. I haven't gone over the notebook to see where this would be used, but if it is related to a question @jessegrabowski asked about covariance matrices in one of the discords you should be able to do something like:

Thanks @OriolAbril, I will give this a try. This would solve an issue that I had originally thought was going to be more complicated than this.

**Comment by Dekermanjian** [comment]
Okay, looks like Plotly is the culprit here. Now I am able to get the math to render properly but now the figures won't render. I'm going to try out a few different renderers to see if I can get both to work at the same time.

**Comment by OriolAbril** [comment]
There seems to be many issues related to plotly/mathjax issues, both on their own and within sphinx builds. Most of them seem to be closed without any clear fix. There are some suggestions regarding mathjax configuration in some but they might be outdated by now. Some examples: https://github.com/jupyter-book/jupyter-book/issues/1528, https://github.com/quarto-dev/quarto-cli/issues/12060, https://github.com/plotly/plotly.js/issues/4563, https://github.com/plotly/plotly.py/issues/3152

**Comment by Dekermanjian** [comment]
Thanks @OriolAbril, I am going to try a handful of renderers and if nothing works I will just turn the figures to static SVG those seem to render just fine.

**Comment by jessegrabowski** [comment]
@OriolAbril do you know how we choose which on the graphs will be used for the thumbnail? It would be great to have one of the maps, they're really attractive.

**Comment by OriolAbril** [comment]
> @OriolAbril do you know how we choose which on the graphs will be used for the thumbnail? It would be great to have one of the maps, they're really attractive.

IIRC, it is the last image in the notebook. It is probablyrestricted to png only though, or maybe png/jpeg only. The code that takes care of that is in https://github.com/pymc-devs/pymc-examples/blob/main/sphinxext/thumbnail_extractor.py. I am quite sure there are other notebooks that add an extra last cell that duplicates the plot they want shown and then [remove that cell from the rendered view](https://myst-nb.readthedocs.io/en/latest/render/hiding.html#remove-parts-of-cells). I think that is significantly easier than adding extra metadata to the notebooks to choose the image and updating the sphinx extension to take that into account.

**Review by jessegrabowski — APPROVED** [review]
: )

**Comment by jessegrabowski** [comment]
@fonnesbeck any objection to merging this as-is?

**Comment by fonnesbeck** [comment]
No objections--it looks great. Merging.
